### PR TITLE
refactor stateful unit, update nletterprocessunit, update unittest

### PR DESCRIPTION
**Pull request by bwanglzu**
For feature #176 
1. Add `NgramLetterProcessorUnit` to `Processors` to generate tri-letters (dependent on `N`).
2. Add unit tests.
3. Refactor unit test for processors.

### CODE REVIEW

File: matchzoo/preprocessors.py [hunk 1]
```diff
@@ -195,6 +195,8 @@ def transform(self, tokens: list, ngram: int=3) -> list:
 
         :param tokens: list of tokens to be transformed.
         :param ngram: By default use 3-gram (tri-letter).
+
+        :return: set of tri-letters, dependent on `ngram`.
```
Comment (faneshion) [inline matchzoo/preprocessors.py]: In the current preprocessors, all operations are conducted on raw texts. However, we have to map each word/n-letters to id, which is  the function of `Vocabulary`.  Thus, we need to design a `Vocabulary` class, whose `fit` construct a _word_dict and `transform` map list of string to list of ids.

Comment (uduse) [reply matchzoo/preprocessors.py]: @faneshion fire up an issue for that?

**Comment by uduse** [comment]
Just dumping more rationale here: `fit` and `transform` should focus on their on jobs. `fit` alters processor unit's inner state without changing input, while `transform` returns a copy of the transformed input without changing inner state. This might be inefficient for some processor units (e.g. n-grams), but it's more important to draw the line between these two things, and performance is not our concern for now. Nonetheless, that inefficiency can be reduced by inducing something like a `fit_transform` in the future, but we don't need that for now.

File: matchzoo/preprocessors.py (hunk 1 above)
Comment (faneshion) [reply matchzoo/preprocessors.py]: OK

**Comment by bwanglzu** [comment]
logically, `fit` should happens before `transform`. While in reality, this doesn't make sense.

If you want to develop a `TrigramProcessUnit` to create `tri-letters` (word -> wor, ord ..), the context (state) you wanna fit is the length of the transformed `tri-letters`. This wouldn't happen before `transform` has been called.

Say if you force to separate `fit` and `transform` operation, then you have two options:
1. Create `tri-letters` again in fit, get it's length of `tri-letters` and throw away the `tri-letters`, computationally expensive.
2. Move `fit` into the follow-up `Process Unit`, then we lose the independence of each units.

**Comment by uduse** [comment]
(1) is exactly what I meant. We do it twice. Quote:

> This might be inefficient for some processor units (e.g. n-grams), but it's more important to draw the line between these two things, and performance is not our concern for now. Nonetheless, that inefficiency can be reduced by inducing something like a fit_transform in the future, but we don't need that for now.

**Comment by bwanglzu** [comment]
This will be inefficient for every `StatefulProcessorUnit`..

**Comment by uduse** [comment]
counter example: [the vocabulary class I just wrote](https://github.com/faneshion/MatchZoo/blob/feature/vocabulary/matchzoo/preprocessors.py#L171)

**Comment by uduse** [comment]
One more thing. The `input` of a `fit` might be different from the `input` of a `transform`, or a number of `transform`s. A common usage is a single `fit` on training data followed by a `transform` on both training and testing data.

File: matchzoo/preprocessors.py [hunk 2]
```diff
@@ -166,3 +166,65 @@ def transform(self, tokens: list) -> list:
...
+    Process unit for n-letter generation.
+
+    Triletter is used in :DSSMModel: and :CDSSMModel:.
+    This processor is expected to execute after `Vocab`
+    has been created.
+
+    Returned `input_dim` is the dimensionality of :DSSMModel:.
+    """
+
+    def __init__(self):
```
Comment (uduse) [inline matchzoo/preprocessors.py]: Not needed.

**Pull request merged**